steam: add extraEnv argument

Atemu opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Atemu committed:
steam: use writeShellScript

Atemu committed:
steam: add extraEnv option

bjornfor commented on nixos/modules/programs/steam.nix:
```diff
@@ -29,6 +30,23 @@ in {
       '';
     };
 
+    env = mkOption {
```
Why does steam need local env vars? Why not use an existing NixOS option to set env vars globally?

Atemu replied:
See the example; these are env vars you might want to set for all games but not system-wide.

Atemu commented:
Bump.

SuperSandro2000 commented on nixos/modules/programs/steam.nix:
```diff
@@ -29,6 +30,23 @@ in {
+      description = mdDoc ''
+        Environment variables to run Steam with.
+
+        Uses `lib.strings.toShellVars` internally; see the documentation for more info.
```
Can we move this to apply and defaultText to make it obvious what happens?

Atemu replied:
I'm not sure what you mean by that.

SuperSandro2000 commented on pkgs/games/steam/fhsenv.nix:
```diff
@@ -249,6 +251,9 @@ in buildFHSUserEnv rec {
     ${exportLDPath}
     ${fixBootstrap}
+
+    set -o allexport # Export the following env vars
```
Should we turn good off afterwards again? I am not sure if it is necessary.

Atemu replied:
The shell script ends there with an exec, so I don't think that's necessary.

Atemu commented:
Bump.

Atemu commented:
If there are no objections, I'll merge this.

Atemu committed:
nixos/steam: always apply extraLibraries and make them additive

Before, setting {option}`programs.steam.package` would result in a steam without
the {option}`hardware.opengl.package`, {option}`hardware.opengl.extraPackages`
etc. You had to manually add them yourself.

Additionally, overlaying `steam = prev.steam.override { extraLibraries = [ ... ]; }`
resulted in those extra libraries not actually being put into the fhsenv because
they'd be fully overridden by the option's default.

Now, the user can supply a custom steam to {option}`programs.steam.package` with
its own list of extraLibraries which will not be overridden and overlays work as
expected too.

Atemu committed:
buildFHSEnvBubblewrap: allow deeper introspection via passthru

Atemu commented:
I've refactored the steam.package option to allow for a much simpler approach than adding a new option.

Atemu commented:
(The first two commits could be spun off to a separate PR.)

RaitoBezarius reviewed (approved):
Untested; but it will probably be tested after merge very well. :)

Atemu commented:
@RaitoBezarius reviewed the new and improved `programs.steam.package` option in the issue above.

I think this is ready to be unleashed upon unstable gamers.
